### PR TITLE
PHP 8.0 support

### LINKED ISSUE #57: PHP 8.0 support
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes

#### Summary

To be prepared for the december release of PHP 8.0, this repository has some additional TODOs to be tested against the new major version.

In order to make this repository compatible, one has to follow these steps:

- [ ] Modify `composer.json` to provide support for PHP 8.0 by adding the constraint `~8.0.0`
- [ ] Modify `composer.json` to drop support for PHP less than 7.3
- [ ] Modify `composer.json` to implement phpunit 9.3 which supports PHP 7.3+
- [ ] Modify `.travis.yml` to ignore platform requirements when installing composer dependencies (simply add `--ignore-platform-reqs` to `COMPOSER_ARGS` env variable)
- [ ] Modify `.travis.yml` to add PHP 8.0 to the matrix (*NOTE*: Do not allow failures as PHP 8.0 has a feature freeze since 2020-08-04!)
- [ ] Modify source code in case there are incompatibilities with PHP 8.0

### PR DESCRIPTION
**Pull request by hummer2k**
|    Q          |   A
|-------------- | ------
| QA            | yes

### Description

This PR adds PHP 8.0 support. 
Phpunit has been updated to version 9.3
Tests have been updated.

See #57

### CODE REVIEW

**Comment by hummer2k** [comment]
@boesing Any idea for the failing composer install? Is the target branch 2.12.x the correct one?

**Comment by rieschl** [comment]
> 
> 
> @boesing Any idea for the failing composer install? Is the target branch 2.12.x the correct one?

@hummer2k You have to add the `COMPOSER_ROOT_VERSION` variable to travis, [like in this PR](https://github.com/laminas/laminas-view/pull/59/commits/d20e7a13ad7b651f2f620f62153ff4f19a42e1a5).
I haven't found a better solution, yet.

**Comment by boesing** [comment]
> @boesing Any idea for the failing composer install? Is the target branch 2.12.x the correct one?



> > @boesing Any idea for the failing composer install? Is the target branch 2.12.x the correct one?
> 
> @hummer2k You have to add the `COMPOSER_ROOT_VERSION` variable to travis, [like in this PR](https://github.com/laminas/laminas-view/pull/59/commits/d20e7a13ad7b651f2f620f62153ff4f19a42e1a5).
> I haven't found a better solution, yet.


For circular-dependencies, its okay to add either the `COMPOSER_ROOT_VERSION` to travis or adding the version to `composer.json` directly.
If the PR is ready to merge, the commit has to get removed (not reverted, removed - so history-rewrite).

This is definitely a thing, not sure how to solve this in a better way tho.

**Comment by rieschl** [comment]
> > @boesing Any idea for the failing composer install? Is the target branch 2.12.x the correct one?
> 
> 
> 
> > > @boesing Any idea for the failing composer install? Is the target branch 2.12.x the correct one?
> > 
> > @hummer2k You have to add the `COMPOSER_ROOT_VERSION` variable to travis, [like in this PR](https://github.com/laminas/laminas-view/pull/59/commits/d20e7a13ad7b651f2f620f62153ff4f19a42e1a5).
> > I haven't found a better solution, yet.
> 
> 
> For circular-dependencies, its okay to add either the `COMPOSER_ROOT_VERSION` to travis or adding the version to `composer.json` directly.
> If the PR is ready to merge, the commit has to get removed (not reverted, removed - so history-rewrite).
> 
> This is definitely a thing, not sure how to solve this in a better way tho.

I _think_ the problem occurred since the move away from `master` and `develop` branch naming. Those branches had a branch alias in `composer.json` so the error didn't occur.
Do why does it hurt if such a declaration is now in travis instead of composer?

**Comment by boesing** [comment]
> I think the problem occurred since the move away from master and develop branch naming. Those branches had a branch alias in composer.json so the error didn't occur.

Nope, its because we dropped `branch-aliases` from the `extra` configuration in `composer.json`.

> Do why does it hurt if such a declaration is now in travis instead of composer?

Because there has to be a commit for *every* new version just to stay in sync. Probably something the release-bot could do while bumping versions in `CHANGELOG` aswell. Maybe @weierophinney and @Ocramius have an opinion about this.

**Comment by rieschl** [comment]
> > I think the problem occurred since the move away from master and develop branch naming. Those branches had a branch alias in composer.json so the error didn't occur.
> 
> Nope, its because we dropped `branch-aliases` from the `extra` configuration in `composer.json`.
> 

Yes, that's what I meant 🙂

> > Do why does it hurt if such a declaration is now in travis instead of composer?
> 
> Because there has to be a commit for *every* new version just to stay in sync. Probably something the release-bot could do while bumping versions in `CHANGELOG` aswell. Maybe @weierophinney and @Ocramius have an opinion about this.

hmm, that's probably a trade-off because the default branch constantly changes 🤷‍♂️

**Comment by boesing** [comment]
> hmm, that's probably a trade-off because the default branch constantly changes 🤷‍♂️

Exactly, thats why I think it should be moved to the release-bot as it creates new release branches and bumps changelog already.

> Yes, that's what I meant 🙂

Sorry, multi-tasking and didn't read all of your message 🙈

File: composer.json [hunk 1]
```diff
@@ -21,12 +21,16 @@
...
     },
     "require": {
-        "php": "^5.6 || ^7.0",
+        "php": "^7.3 || ~8.0.0",
         "laminas/laminas-eventmanager": "^2.6.2 || ^3.0",
         "laminas/laminas-json": "^2.6.1 || ^3.0",
         "laminas/laminas-loader": "^2.5",
+        "laminas/laminas-mvc-console": "^1.2",
+        "laminas/laminas-mvc-i18n": "^1.1",
+        "laminas/laminas-mvc-plugin-flashmessenger": "^1.2",
```
Comment (froschdesign) [inline composer.json]: Please remove these packages as they are not required.

File: composer.json [hunk 2]
```diff
@@ -21,12 +21,16 @@
...
         "laminas/laminas-stdlib": "^2.7 || ^3.0",
-        "laminas/laminas-zendframework-bridge": "^1.0"
+        "laminas/laminas-zendframework-bridge": "^1.0",
+        "phpspec/prophecy-phpunit": "^2.0"
```
Comment (froschdesign) [inline composer.json]: Wrong section. Please move this package to the `require-dev` section. Thanks!

**Comment by hummer2k** [comment]
Okay, travis is now green.

What's next?

- Drop Service-Manager v2 and related tests?
- Drop further v2 components?
- Any suggestions regarding the deprecated xml assertions? (https://travis-ci.com/github/laminas/laminas-view/jobs/408013304#L566)

**Comment by froschdesign** [comment]
@hummer2k 
> Any suggestions regarding the deprecated xml assertions? (https://travis-ci.com/github/laminas/laminas-view/jobs/408013304#L566)

One option in this case is to compare the strings:

```php
self::assertEquals($expectedDom->saveXML(), $receivedDom->saveXML());
```

**Comment by boesing** [comment]
@hummer2k huge addition here! Thanks!
I'd suggest dropping eventmanager and servicemanager v2 support. 
If there are tests related to the v2 versions of those packages, feel free to drop these.
If needed, adding `conflict` for servicemanager <3.x should be fine.

**Review by boesing — APPROVED** [review]
LGTM!

**Comment by boesing** [comment]
Thanks, @hummer2k!

**Pull request merged**